Add spec for persist module (#109)
ballerina-platform/module-ballerina-persist: docs/spec/spec.md

### 3.1. Derived Entity Types

Ballerina record types used in the `PersistClient` are derived from the entity types defined in the data model.

Example of a defined entity,
```ballerina
type Workspace record {|
    readonly string workspaceId;
    string workspaceType;
    Building location; // 1-n relation (parent)
    Employee employee; // 1-1 relation (child)
|};
```

There are 3 types of derived entity types:
1. Entity Types  
    These are the types defined in the data model. These are used to define the structure of the data in the data source. 
    The entity who is the parent of the association will include foreign key field.
    ```ballerina
    public type Workspace record {|
        readonly string workspaceId;
        string workspaceType;
        string buildingBuildingCode;
    |};
    ```
    
2. Insert Types
    These are records used to insert data in the data source. It is the same as the entity type.
    ```ballerina
    type WorkspaceInsert Workspace;
    ```

3. Update Types
    These are records used to update data in the data source. These are entity types without the identifier fields.
    ```ballerina

>     type WorkspaceUpdate record {|

Reviewer: shouldn't these be nullable types, as in [here](https://github.com/ballerina-platform/module-ballerina-persist/blob/005e6fafd3a21fec030880da856a604728eac401/ballerina/tests/generated_types.bal#L58)

Author: What you mean is its optional field? `int field?;` ?

Reviewer: Sorry it should be `optional field`

> # Specification: Ballerina SQL Library

Reviewer 2: ```suggestion
# Specification: Ballerina Persist Library
```

_Owners_: @daneshk @niveathika @kaneeldias
_Reviewers_: @daneshk  
_Created_: 2022/01/25   
_Updated_: 2022/01/25  
_Edition_: Swan Lake  

## Introduction

> This is the specification for the Ballerina persist standard library of [Ballerina language](https://ballerina.io/), which provides functionality to define and store data.  

Reviewer 2: ```suggestion
This is the specification for the Ballerina persist standard library of [Ballerina language](https://ballerina.io/), which provides functionality to store and query data easily.  
```

The conforming implementation of the specification is released and included in the distribution. Any deviation from the specification is considered a bug.

## Contents

1. [Overview](#1-overview)  
2. [Data Model Definition](#2-data-model-definition)  
    2.1. [EntityType Definition](#21-entitytype-definition)  
    2.2. [Entity Attributes Definition](#22-entity-attributes-definition)  
        2.2.1. [Identifier Field(s)](#221-identifier-fields)  
        2.2.2. [Nullable Field(s)](#222-nullable-fields)  
    2.3. [Relationship Definition](#23-relationship-definition)    
        2.3.1. [One-to-one (1-1)](#231-one-to-one-1-1)  
        2.3.2. [One-to-Many (1-n)](#232-one-to-many-1-n)  
3. [Derived Entity Types and Persist Clients](#3-derived-entity-types-and-persist-clients)  
    3.1. [Derived Entity Types](#31-derived-entity-types)  
    3.2. [Persist Clients](#32-persist-clients)

## 1. Overview  

> The `persist` standard library creates a programming model to store and query data easily. This uses the statically typed nature of Ballerina to provide a convenient way to define and store data model of an application across various providers such as RDBMS, NoSQL, and others. At this initial release `persist` library only supports MySQL data provider.

Reviewer 2: ```suggestion
The `persist` standard library creates a programming model to store and query data easily. This uses the statically typed nature of Ballerina to provide a convenient way to define and store the data model of an application across various providers such as RDBMS, NoSQL, and others. At this initial release `persist` library only supports MySQL data provider.
```

    ```ballerina
    type WorkspaceUpdate record {|
        string workspaceType?;
        string buildingBuildingCode?;
    |};
    ```

### 3.2. Persist Clients

Persist Clients are derived for each data model definition file.

The skeleton of the Persist client is as follows,
```ballerina
client class RainierClient {
    *persist:AbstractPersistClient;

    public function init() returns persist:Error? {
    }

    isolated resource function get workspaces() returns stream<Workspace, persist:Error?> {
    };

    isolated resource function get workspaces/[string workspaceId]() returns Workspace|persist:Error {
    };

    isolated resource function post workspaces(WorkspaceInsert[] data) returns string[]|persist:Error {
    };

    isolated resource function put workspaces/[string workspaceId](WorkspaceUpdate data) returns Workspace|persist:Error {
    };

    isolated resource function delete workspaces/[string workspaceId]() returns Workspace|persist:Error {
    };

    public function close() returns persist:Error? {
    }
}
```

The conventions used in deriving the Persist client are as follows:

> 1.  The Client name is derived from the name of the file. Example: `rainier.bal` will generate a client named `RainierClient`(<First Letter Capitalized File Name>Client).
> 2. The client should be of `persist:AbstractPersistClient` type.

Reviewer 2: ```suggestion
1. The Client name is derived from the name of the file. Example: `rainier.bal` will generate a client named `RainierClient`(<First Letter Capitalized File Name>Client).
2. The client should be of `persist:AbstractPersistClient` type.
```

## Introduction

> This is the specification for the Ballerina persist standard library of [Ballerina language](https://ballerina.io/), which provides functionality to store and query data easily.

Reviewer 3: ```suggestion
This is the specification for the `persist` standard library of the [Ballerina language](https://ballerina.io/), which provides functionality to store and query data easily.
```

Author: Fixed in [e4dc132](https://github.com/ballerina-platform/module-ballerina-persist/pull/109/commits/e4dc1324152adf53e52a1ca4dea90ffacb44ae7a)

> The persist library specification has evolved and may continue to evolve in the future. The released versions of the specification can be found under the relevant GitHub tag. 

Reviewer 3: ```suggestion
The `persist` library specification has evolved and may continue to evolve in the future. The released versions of the specification can be found under the relevant GitHub tag. 
```
Please check for the backticks in all occurrences of the module name.

        2.3.2. [One-to-Many (1-n)](#232-one-to-many-1-n)  
3. [Derived Entity Types and Persist Clients](#3-derived-entity-types-and-persist-clients)  
    3.1. [Derived Entity Types](#31-derived-entity-types)  
    3.2. [Persist Clients](#32-persist-clients)  

## 1. Overview

> The `persist` standard library creates a programming model to store and query data easily. This uses the statically typed nature of Ballerina to provide a convenient way to define and store the data model of an application across various providers such as RDBMS, NoSQL, and others. At this initial release `persist` library only supports MySQL data provider.

Reviewer 3: ```suggestion
The `persist` standard library creates a programming model to store and query data easily. This uses the statically typed nature of Ballerina to provide a convenient way to define and store the data model of an application across various providers such as RDBMS, NoSQL, and others. At this initial release, the `persist` library only supports the MySQL data provider.
```

Author: Fixed in [e4dc132](https://github.com/ballerina-platform/module-ballerina-persist/pull/109/commits/e4dc1324152adf53e52a1ca4dea90ffacb44ae7a)
